File: samples/C# Samples/DotNETTWAINDemo/DotNETTWAINDemo/ResampleForm.cs
using Dynamsoft.Core.Enums;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;

namespace DotNet_TWAIN_Demo
{
    public partial class ResampleForm : Form
    {
        private int width, height;
        private int newWidth, newHeight;
        private bool flag;
        private EnumInterpolationMethod m_InterpolationMethod = EnumInterpolationMethod.BestQuality;
        public ResampleForm()
        {
            InitializeComponent();
        }

        public ResampleForm(int width, int height)
        {
            InitializeComponent();
            
            this.width = width;
            this.height = height;

            this.flag = false;
            this.tbxWidth.Text = width.ToString();
            this.tbxHeight.Text = height.ToString();
            this.cbxWidthType.SelectedIndex = 0;
            this.cbxHeightType.SelectedIndex = 0;
            this.cbxConstrainProportion.Checked = true;
            this.cbxResampleType.SelectedIndex = 0;
            flag = true;
        }
        
        public int NewHeight
        {
            get 
            {
                if (cbxWidthType.SelectedIndex == 0)
                    return int.Parse(tbxHeight.Text);
                else
                    return int.Parse(tbxHeight.Text) * height / 100;
            }
        }

        public int NewWidth
        {
            get
            {
                if (cbxHeightType.SelectedIndex == 0)
                    return int.Parse(tbxWidth.Text);
                else
                    return int.Parse(tbxWidth.Text) * width / 100;
            }
        }

        public EnumInterpolationMethod Interpolation
        {
            get
            {
                switch (this.cbxResampleType.SelectedIndex)
                {
                    case 0:
                        m_InterpolationMethod = EnumInterpolationMethod.Bicubic;
                        break;
                    case 1:
                        m_InterpolationMethod = EnumInterpolationMethod.Bilinear;
                        break;
                    case 2:
                        m_InterpolationMethod = EnumInterpolationMethod.NearestNeighbour;
                        break;
                    case 3:
                        m_InterpolationMethod = EnumInterpolationMethod.BestQuality;
                        break;
                }
                return m_InterpolationMethod;
            }
        }
        private void btnOk_Click(object sender, EventArgs e)
        {
            if (!int.TryParse(tbxWidth.Text, out newWidth))
            {
                tbxWidth.Focus();
                return;
            }
            if (!int.TryParse(tbxHeight.Text, out newHeight))
            {
                tbxHeight.Focus();
                return;
            }
            this.DialogResult = DialogResult.OK;
            this.Close();
        }

        private void cbxWidthType_SelectedIndexChanged(object sender, EventArgs e)
        {
            if (this.cbxHeightType.SelectedIndex != this.cbxWidthType.SelectedIndex && flag)
            {
                if (cbxWidthType.SelectedIndex == 0)
                {
                    if (int.TryParse(tbxWidth.Text, out newWidth) && int.TryParse(tbxHeight.Text, out newHeight))
                    {
                        newHeight = newHeight * height / 100;
                        newWidth = newWidth * width / 100;
                        tbxWidth.Text = newWidth.ToString();
                        tbxHeight.Text = newHeight.ToString();
                    }
                }
                else
                {
                    if (int.TryParse(tbxWidth.Text, out newWidth) && int.TryParse(tbxHeight.Text, out newHeight))
                    {
                        newWidth = (int)(newWidth / (double)width * 100);
                        newHeight = (int)(newHeight / (double)height * 100);
                        tbxWidth.Text = newWidth.ToString();
                        tbxHeight.Text = newHeight.ToString();
                    }
                }
                
                this.cbxHeightType.SelectedIndex = this.cbxWidthType.SelectedIndex;
            }
        }

        private void cbxHeightType_SelectedIndexChanged(object sender, EventArgs e)
        {
            if (this.cbxWidthType.SelectedIndex != this.cbxHeightType.SelectedIndex && flag)
            {
                if (cbxHeightType.SelectedIndex == 0)
                {
                    if (int.TryParse(tbxHeight.Text, out newHeight) && int.TryParse(tbxWidth.Text, out newWidth))
                    {
                        newWidth = newWidth * width / 100;
                        newHeight = newHeight * height / 100;
                        tbxWidth.Text = newWidth.ToString();
                        tbxHeight.Text = newHeight.ToString();
                    }
                }
                else
                {
                    if (int.TryParse(tbxHeight.Text, out newHeight))
                    {
                        newHeight = (int)(newHeight / (double)height * 100);
                        newWidth = (int)(newWidth / (double)width * 100);
                        tbxWidth.Text = newWidth.ToString();
                        tbxHeight.Text = newHeight.ToString();
                    }
                }
                this.cbxWidthType.SelectedIndex = this.cbxHeightType.SelectedIndex;
            }
        }

        private void tbxWidth_KeyUp(object sender, KeyEventArgs e)
        {
            if (cbxConstrainProportion.Checked == true)
            {
                if (int.TryParse(tbxWidth.Text, out newWidth) == true)
                {
                    double proportion = width / (double)height;
                    if (cbxWidthType.SelectedIndex == 0)
                    {
                        tbxHeight.Text = ((int)(newWidth / proportion)).ToString();
                    }
                    if (cbxWidthType.SelectedIndex == 1)
                    {
                        tbxHeight.Text = newWidth.ToString();
                    }
                }
            }
        }

        private void tbxHeight_KeyUp(object sender, KeyEventArgs e)
        {
            if (cbxConstrainProportion.Checked == true)
            {
                if (int.TryParse(tbxHeight.Text, out newHeight) == true)
                {
                    double proportion = width / (double)height;
                    if (cbxWidthType.SelectedIndex == 0)
                    {
                        tbxWidth.Text = ((int)(newHeight * proportion)).ToString();
                    }
                    if (cbxWidthType.SelectedIndex == 1)
                    {
                        tbxWidth.Text = newHeight.ToString();
                    }
                }
            }
        }

        private void btnCancel_Click(object sender, EventArgs e)
        {
            this.DialogResult = DialogResult.Cancel;
            this.Close();
        }
    }
}

File: samples/VB .NET Samples/DotNETTWAINDemo/DotNETTWAINDemo/ResampleForm.vb
Imports Dynamsoft.Core.Enums
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms

Public Partial Class ResampleForm
	Inherits Form
	Private width As Integer, height As Integer
	Private m_newWidth As Integer, m_newHeight As Integer
	Private flag As Boolean
	Private m_InterpolationMethod As EnumInterpolationMethod = EnumInterpolationMethod.BestQuality
	Public Sub New()
		InitializeComponent()
	End Sub

	Public Sub New(width As Integer, height As Integer)
		InitializeComponent()

		Me.width = width
		Me.height = height

		Me.flag = False
		Me.tbxWidth.Text = width.ToString()
		Me.tbxHeight.Text = height.ToString()
		Me.cbxWidthType.SelectedIndex = 0
		Me.cbxHeightType.SelectedIndex = 0
		Me.cbxConstrainProportion.Checked = True
		Me.cbxResampleType.SelectedIndex = 0
		flag = True
	End Sub

	Public ReadOnly Property NewHeight() As Integer
		Get
			If cbxWidthType.SelectedIndex = 0 Then
				Return Integer.Parse(tbxHeight.Text)
			Else
				Return Integer.Parse(tbxHeight.Text) * height \ 100
			End If
		End Get
	End Property

	Public ReadOnly Property NewWidth() As Integer
		Get
			If cbxHeightType.SelectedIndex = 0 Then
				Return Integer.Parse(tbxWidth.Text)
			Else
				Return Integer.Parse(tbxWidth.Text) * width \ 100
			End If
		End Get
	End Property

	Public ReadOnly Property Interpolation() As EnumInterpolationMethod
		Get
			Select Case Me.cbxResampleType.SelectedIndex
				Case 0
					m_InterpolationMethod = EnumInterpolationMethod.Bicubic
					Exit Select
				Case 1
					m_InterpolationMethod = EnumInterpolationMethod.Bilinear
					Exit Select
				Case 2
					m_InterpolationMethod = EnumInterpolationMethod.NearestNeighbour
					Exit Select
				Case 3
					m_InterpolationMethod = EnumInterpolationMethod.BestQuality
					Exit Select
			End Select
			Return m_InterpolationMethod
		End Get
	End Property
	Private Sub btnOk_Click(sender As Object, e As EventArgs)
		If Not Integer.TryParse(tbxWidth.Text, m_newWidth) Then
			tbxWidth.Focus()
			Return
		End If
		If Not Integer.TryParse(tbxHeight.Text, m_newHeight) Then
			tbxHeight.Focus()
			Return
		End If
		Me.DialogResult = DialogResult.OK
		Me.Close()
	End Sub

	Private Sub cbxWidthType_SelectedIndexChanged(sender As Object, e As EventArgs)
		If Me.cbxHeightType.SelectedIndex <> Me.cbxWidthType.SelectedIndex AndAlso flag Then
			If cbxWidthType.SelectedIndex = 0 Then
				If Integer.TryParse(tbxWidth.Text, m_newWidth) AndAlso Integer.TryParse(tbxHeight.Text, m_newHeight) Then
					m_newHeight = m_newHeight * height \ 100
					m_newWidth = m_newWidth * width \ 100
					tbxWidth.Text = m_newWidth.ToString()
					tbxHeight.Text = m_newHeight.ToString()
				End If
			Else
				If Integer.TryParse(tbxWidth.Text, m_newWidth) AndAlso Integer.TryParse(tbxHeight.Text, m_newHeight) Then
					m_newWidth = CInt(Math.Truncate(m_newWidth / CDbl(width) * 100))
					m_newHeight = CInt(Math.Truncate(m_newHeight / CDbl(height) * 100))
					tbxWidth.Text = m_newWidth.ToString()
					tbxHeight.Text = m_newHeight.ToString()
				End If
			End If

			Me.cbxHeightType.SelectedIndex = Me.cbxWidthType.SelectedIndex
		End If
	End Sub

	Private Sub cbxHeightType_SelectedIndexChanged(sender As Object, e As EventArgs)
		If Me.cbxWidthType.SelectedIndex <> Me.cbxHeightType.SelectedIndex AndAlso flag Then
			If cbxHeightType.SelectedIndex = 0 Then
				If Integer.TryParse(tbxHeight.Text, m_newHeight) AndAlso Integer.TryParse(tbxWidth.Text, m_newWidth) Then
					m_newWidth = m_newWidth * width \ 100
					m_newHeight = m_newHeight * height \ 100
					tbxWidth.Text = m_newWidth.ToString()
					tbxHeight.Text = m_newHeight.ToString()
				End If
			Else
				If Integer.TryParse(tbxHeight.Text, m_newHeight) Then
					m_newHeight = CInt(Math.Truncate(m_newHeight / CDbl(height) * 100))
					m_newWidth = CInt(Math.Truncate(m_newWidth / CDbl(width) * 100))
					tbxWidth.Text = m_newWidth.ToString()
					tbxHeight.Text = m_newHeight.ToString()
				End If
			End If
			Me.cbxWidthType.SelectedIndex = Me.cbxHeightType.SelectedIndex
		End If
	End Sub

	Private Sub tbxWidth_KeyUp(sender As Object, e As KeyEventArgs)
		If cbxConstrainProportion.Checked = True Then
			If Integer.TryParse(tbxWidth.Text, m_newWidth) = True Then
				Dim proportion As Double = width / CDbl(height)
				If cbxWidthType.SelectedIndex = 0 Then
					tbxHeight.Text = CInt(Math.Truncate(m_newWidth / proportion)).ToString()
				End If
				If cbxWidthType.SelectedIndex = 1 Then
					tbxHeight.Text = m_newWidth.ToString()
				End If
			End If
		End If
	End Sub

	Private Sub tbxHeight_KeyUp(sender As Object, e As KeyEventArgs)
		If cbxConstrainProportion.Checked = True Then
			If Integer.TryParse(tbxHeight.Text, m_newHeight) = True Then
				Dim proportion As Double = width / CDbl(height)
				If cbxWidthType.SelectedIndex = 0 Then
					tbxWidth.Text = CInt(Math.Truncate(m_newHeight * proportion)).ToString()
				End If
				If cbxWidthType.SelectedIndex = 1 Then
					tbxWidth.Text = m_newHeight.ToString()
				End If
			End If
		End If
	End Sub

	Private Sub btnCancel_Click(sender As Object, e As EventArgs)
		Me.DialogResult = DialogResult.Cancel
		Me.Close()
	End Sub
End Class
